Clear general cache less often when launching the archiving

tsteur opened the pull request:
### Description:

fix DEV-2707

Since Matomo 4.7 we noticed that we often have thousands of general cache invalidations within a minute. Around 100 general cache invalidations within approx 4 seconds but depends on how many archives there are to be invalidated. It causes one general cache invalidation per site every time we call `prepareArchive`. Aka if you have thousands of sites, and run heaps of archives, and all sites have regular traffic, this can cause a huge massive amount of invalidations.

<img width="175" alt="image" src="https://user-images.githubusercontent.com/273120/160057473-791dbb59-6345-4520-8796-c1fbb456e987.png">


All these cache invalidations are also particularly a problem as we're constantly clearing the cache, and also constantly populate the tracker cache every ms as there are concurrent tracking requests happening. 

The problem is coming from this backtrace which is called once per site when preparing an archive run. So this can be called very often this logic and cause a huge amount of invalidations.

```

#0  Piwik\Tracker\Cache::clearCacheGeneral() called at [core/Archive/ArchiveInvalidator.php:332]
#1  Piwik\Archive\ArchiveInvalidator->markArchivesAsInvalidated() called at [plugins/CoreAdminHome/API.php:166]
#2  Piwik\Plugins\CoreAdminHome\API->invalidateArchivedReports() called at [core/CronArchive.php:928]
#3  Piwik\CronArchive->invalidateWithSegments() called at [core/CronArchive.php:873]
#4  Piwik\CronArchive->invalidateArchivedReportsForSitesThatNeedToBeArchivedAgainImpl() called at [core/CronArchive.php:808]
#5  Piwik\CronArchive->Piwik\{closure}() called at [core/Tracker/Cache.php:302]
#6  Piwik\Tracker\Cache::withDelegatedCacheClears() called at [core/CronArchive.php:809]
#7  Piwik\CronArchive->invalidateArchivedReportsForSitesThatNeedToBeArchivedAgain() called at [core/CronArchive/QueueConsumer.php:182]
#8  Piwik\CronArchive\QueueConsumer->getNextArchivesToProcess() called at [core/CronArchive.php:394]
#9  Piwik\CronArchive->run() called at [core/CronArchive.php:280]
#10 Piwik\CronArchive->Piwik\{closure}() called at [core/Access.php:661]
#11 Piwik\Access::doAsSuperUser() called at [core/CronArchive.php:286]
#12 Piwik\CronArchive->main() called at [plugins/CoreConsole/Commands/CoreArchiver.php:32]
#13 Piwik\Plugins\CoreConsole\Commands\CoreArchiver->execute() called at [vendor/symfony/console/Symfony/Component/Console/Command/Command.php:257]
#14 Symfony\Component\Console\Command\Command->run() called at [vendor/symfony/console/Symfony/Component/Console/Application.php:874]
#15 Symfony\Component\Console\Application->doRunCommand() called at [vendor/symfony/console/Symfony/Component/Console/Application.php:195]
#16 Symfony\Component\Console\Application->doRun()
```

To improve the situation my thought was to invalidate the general cache only every 4 seconds. We could technically only clear the cache after marking all archives as invalid. However, this could take a while depending on the number of archives to invalidate and the number of sites. 

If it takes a few minutes between marking an archive as invalid and calling `clearGeneralCache`, then we risk that the given date won't be re-archived later under circumstances since the tracker wouldn't know we have invalidated the archive and therefore it wouldn't call `rememberToInvalidateArchivedReportsLater` and then we wouldn't archive that date again under some circumstances. It's bit hard to explain but generally we need a general cache clear ideally immediately after marking an archiving as invalid.

My trade-off was that it's unlikely that the only tracking request for the given date happens during those 5 seconds. And for today periods we always archive the date again the next day after midnight so it shouldn't be an issue.

Any thoughts on this? I haven't fully tested it yet but wanted to check for thoughts first.

There could be other solutions like having the information to `rememberToInvalidateArchivedReportsLater` in a tracker site cache but thinking this can make things worse as we can't optimise it as well as this one and causes again a lot of invalidations.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

tsteur committed:
Clear general cache less often when launching the archiving

sgiehl commented:
I actually wasn't aware of that special caching mechanism. But I had a closer look. As far as I have seen those options are actually only stored to the general cache, as the tracker uses the value to check if the archives need to be invalidated or not. If I did not miss anything that is actually only done for tracking requests with a date before today. See https://github.com/matomo-org/matomo/blob/3b00013819863472afe30cc04de49c4365c22d21/core/Tracker/Visit.php#L622-L628
*Note: Maybe we could even change that to before yesterday, as we always rearchive yesterday nevertheless, don't we?*

So for most websites tracking older visits is very unlikely. The cache of those options seems only to be used within the Tracker. So regarding clearing the cache in `markArchivesAsInvalidated`, I was wondering if we can't simply skip it if another period than day should be invalidated or if the invalidated date is today (or yesterday). As in that cases updating the cache won't have any effect at all.

Your solution could be used on top maybe to avoid too many invalidations when older dates are invalidated.
Hope that makes sense...

tsteur committed:
invalidate only when needed

tsteur committed:
optimise clear general cache

tsteur committed:
better naming

tsteur commented:
That's a great point @sgiehl 💯  This should reduce the general cache invalidations already quite a bit possibly (although not sure we set the flag for today or not).

I've adjusted the logic to
 
* Only clear general cache if the date is not for today (using the [same logic as in ArchiveInvalidator](https://github.com/matomo-org/matomo/blob/4.8.0/core/Archive/ArchiveInvalidator.php#L280-L282))
* And only clear cache every 4 seconds only if any of the previous dates actually needed an invalidation

The logic is bit complicated to understand maybe but should tweak the performance well. I haven't fully tested it yet but wanted to see if you have any thoughts on this first @sgiehl ?

sgiehl commented:
@tsteur I spent some more thoughts on that. Actually I think we did that a lot more complex than actually needed.
Clearing the cache is actually only needed if an option for a remembered invalidation was removed from the database.
That is triggered here:
https://github.com/matomo-org/matomo/blob/5a3ae2d7ff17d1d6bd11ffa65f78101c296bd99e/core/Archive/ArchiveInvalidator.php#L318-L331

So if we would change the `forgetRememberedArchivedReportsToInvalidate` to return true if a record has been removed, we could only clear the cache if that happened at least once within the foreach.
That should limit the number of cache clears to the number of option records in the database.

tsteur committed:
try better approach

tsteur committed:
simplify

tsteur committed:
simplify further

tsteur commented:
@sgiehl great point, I didn't notice 🚀  that made it a lot easier. 

For now I also removed the code that would invalidate only every 4 seconds as it be hard to implement with the deleted flag. I would do it this way, and then see if it's still a problem or not. If it then is still a problem, would add some extra code to invalidate less often.

Because in https://github.com/matomo-org/matomo/blob/3b00013819863472afe30cc04de49c4365c22d21/core/Tracker/Visit.php#L622-L628 which you pointed out we don't invalidate for today anyway, I also removed the check for `isToday` which wasn't needed in the first place.

sgiehl reviewed (approved):
Looks good to me now.
@tsteur should this one go into the upcoming 4.9 or shall we wait for 4.10?

tsteur commented:
@sgiehl I'll leave this up to you. Safest be probably to put it into 4.10 and we could always apply a patch for this on production.

sgiehl commented:
ok. moved it to 4.10..
